Spaces - Copy Saved Objects to Spaces UI

legrego opened the pull request:
## Summary

This introduces a new "Copy to space" action within the Saved Objects Management UI. When Spaces is enabled, a new action appears in the table next to each saved object. From there, users are given the option to copy that saved object (with or without references) to one or more other spaces.

## Tasks
- [x] create workflow
- [x] create integration points in saved objects management ui
- [x] integrate with actual APIs from #38014 
- [x] testing
- [x] i18n verification
- [x] a11y verification
- [ ] copy review


![cts_demo](https://user-images.githubusercontent.com/3493255/62297796-d3270580-b43f-11e9-9b83-bd0193188991.gif)



### Menu option
![cts_1](https://user-images.githubusercontent.com/3493255/62297826-e2a64e80-b43f-11e9-91bb-55995066bb38.png)

### Select spaces to copy into
![cts_2](https://user-images.githubusercontent.com/3493255/62297867-ec2fb680-b43f-11e9-9076-ed87dbb27487.png)

### Processing copy
![cts_3](https://user-images.githubusercontent.com/3493255/62297879-f3ef5b00-b43f-11e9-880e-4f1b0e677b22.png)


### Copy response, showing sample error
![cts_4](https://user-images.githubusercontent.com/3493255/62297892-fd78c300-b43f-11e9-8b91-bfc611ddb499.png)

### Copy response, showing conflict resolution
![cts_5](https://user-images.githubusercontent.com/3493255/62297927-09fd1b80-b440-11e9-90b8-d5b450ef68a1.png)

![cts_6](https://user-images.githubusercontent.com/3493255/62297944-12555680-b440-11e9-8223-1f8f3a469604.png)


Resolves #37286


"Release notes: Saved objects can be easily copied from one space to another from the Saved Objects Management screen"

legrego commented:
@cchaos no rush, as the implementation is blocked on several other PRs, but if you get some free time, I'd love a preliminary design review!

elasticmachine commented:
Pinging @elastic/kibana-security

legrego commented on src/legacy/ui/public/management/saved_objects_management/saved_objects_management_action.ts:
```diff
@@ -0,0 +1,68 @@
+  };
+}
+
+export abstract class SavedObjectsManagementAction<T = unknown> {
```
This is tightly coupled to both EUI and React. My ultimate goal is to replace this entirely with the Embeddable/Actions APIs once they're ready, so I don't want to spend a ton of time crafting the perfect abstraction here. Instead, this provides enough functionality to enable the copy-to-space feature, without coupling it to Spaces itself.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2156/)

legrego commented on src/legacy/ui/public/management/saved_objects_management/saved_objects_management_action.ts:
```diff
@@ -0,0 +1,68 @@
+    render?: (item: SavedObjectRecord) => any;
+  };
+
+  private isActive: boolean = false;
```
TODO for myself: remove `isActive`, we aren't using this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3853/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5603/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6165/)

Conky5 commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6192/)

cchaos commented:
@legrego Sorry it's taken so long for the design team to take a look at this. @ryankeairns will actually be able to help you out on it. You'll probably want to sync with him to go over the purpose and intricacies of all the Spaces stuffs.

legrego commented:
@ryankeairns I updated the screenshots in this PR summary to reflect my changes based on your mockup. It's not production ready, but it hopefully gives you an idea of the direction I took.

I deviated from your mockup in a couple of places. I'm more than willing to make any adjustments you see fit:
- Added "skip" option to undo the "overwrite" button click. Based on the mockup, there was no way to indicate that you no longer wanted to overwrite the saved object. This adds a "skip" option if you want to cancel this operation.
- Added selected options to the top of the processing screen, which just reflects the chosen options (include related, automatically overwrite)
- Added operation summary to the footer, showing number of copied, pending, skipped (due to conflict), and errors.

Since there is no way to automatically expand accordions (except on initial render), I wanted a way to surface the fact that something might need your attention, so I put that summary near the "Finish" button to hopefully make that a little clearer. 


This also includes [a couple of sass entries](https://github.com/legrego/kibana/blob/copy-to-space-ui/x-pack/legacy/plugins/spaces/public/views/management/components/copy_to_space/_index.scss), which are probably worth looking at to make sure I'm not doing anything crazy.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10428/)

ryankeairns commented:
@legrego This is looking really great, thanks for making those changes from the mockups.

Here is a design PR that contains a smallish set of touchups that mostly address alignment type things, spacing, etc. Merge at your leisure!

#### Design PR
👁 👉 [https://github.com/legrego/kibana/pull/22](https://github.com/legrego/kibana/pull/22)

legrego commented:
@ryankeairns thanks to you it looks even better - I appreciate the design edits!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10603/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12221/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12661/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12889/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13553/)

legrego commented:
@kobelb this is generally ready for review. Waiting to see if I have a clean CI run yet, but otherwise we're in good shape

legrego commented:
@elastic/kibana-design you've been flagged for review, but we haven't made any sass/style changes since @ryankeairns [reviewed and adjusted](https://github.com/elastic/kibana/pull/39002#issuecomment-517532826).

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13824/)

spalger commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13872/)

kobelb commented on x-pack/legacy/plugins/spaces/public/lib/copy_saved_objects_to_space/summarize_copy_result.ts:
```diff
@@ -0,0 +1,142 @@
+    conflicts: conflicts.filter(
+      c => c.obj.type === savedObject.type && c.obj.id === savedObject.id
+    ),
+    missingReferences,
```
Currently, if you try to use `export` or `copy to space` where a reference is missing, we get a `400` Bad Request is thrown. So, it's hard to tell how the missing references will actually be returned once this is fixed.

Do we really want all missing reference conflicts to show up here? Wouldn't this be all missing references for the entire "object graph" including transitive references? 

For the direct references and transitive references, we end up listing all missing reference errors that are for the object itself. If that's correct, shouldn't the "missing reference" for the saved object we're trying to copy always be `null`/`undefined`?

legrego replied:
Hmm, I think we're talking about different scenarios here. Any `copy to space` code that mentions "missing references" is talking about conditions which produce this **import** error: https://github.com/elastic/kibana/blob/4996c5d62e745e07b746c12186d65195be270c84/src/core/server/saved_objects/import/types.ts#L68-L78

Specifically, there are two conditions which cause this (that I'm aware of), both of which are produced by [`validate_references`](https://github.com/elastic/kibana/blob/e137477b5a00b8c83e63acda1121c1d2ee10541c/src/core/server/saved_objects/import/validate_references.ts#L79):
- An object being imported directly depends on an index pattern which isn't provided in the upload/copy request, nor in the target space (e.g., unchecking "include references" when copying a visualization which references an index pattern into a fresh space)
- An object being imported directly depends on a saved search which isn't provided in the upload/copy request, nor in the target space (e.g., unchecking "include references" when copying an index pattern which references a saved search into a fresh space)

The normal import/export flow shows a screen which allows users to choose a new index pattern or saved search to associate, but this is one of the features that differs between `import/export` and `copy to space`.

I have this "missing references" code in place for the sole purpose of [showing a more informative tooltip](https://github.com/elastic/kibana/blob/139d606215138e5f076248d0d29e9d6df8405e8b/x-pack/legacy/plugins/spaces/public/views/management/components/copy_saved_objects_to_space/copy_status_indicator.tsx#L67) when reviewing the "copy to space" results.


> Wouldn't this be all missing references for the entire "object graph" including transitive references?

No, currently, the import/export service only deals with direct references to index patterns and saved searches. No other reference types are checked for this condition on import, and we don't fail if this happens on a transitive reference either. This feels inconsistent to me, but also outside the scope of the "Copy to Space" feature IMO.

kobelb commented on x-pack/legacy/plugins/spaces/public/views/management/components/copy_saved_objects_to_space/copy_status_indicator.tsx:
```diff
@@ -0,0 +1,116 @@
+    );
+    return <EuiIconTip type={'check'} color={successColor} content={message} />;
+  }
+  if (hasMissingRefs) {
```
Were you able to test this functionality somehow? I get a 400 whenever trying...

legrego replied:
See [comment above](https://github.com/elastic/kibana/pull/39002#discussion_r316951202), I think we're talking about two different scenarios

kobelb commented on x-pack/legacy/plugins/spaces/public/lib/copy_saved_objects_to_space/summarize_copy_result.ts:
```diff
@@ -0,0 +1,142 @@
+  | UnsuccessfulResponse
+  | ProcessingResponse;
+
+export function summarizeCopyResult(
```
If we only use this from within `x-pack/legacy/plugins/spaces/public/views/management/components/copy_saved_objects_to_space` should we move it there?

legrego replied:
We can...I chose to put it here so that I could test it independently from the React components, and it felt wrong to have this non-component file within the `**/components/**` directory structure

legrego commented on src/legacy/ui/public/management/saved_objects_management/index.ts:
```diff
@@ -0,0 +1,29 @@
+export {
+  processImportResponse,
+  ProcessedImportResponse,
+} from '../../../../core_plugins/kibana/public/management/sections/objects/lib/process_import_response';
```
Ah, I thought this was acceptable because we do the same thing from [`src/legacy/ui/public/management/index.js`](https://github.com/elastic/kibana/blob/2ce51a5be583a0a58c38ed1426efdbb9e5f19ecc/src/legacy/ui/public/management/index.js#L20-L28) today to share functionality between OSS and X-Pack

legrego commented on x-pack/legacy/plugins/spaces/public/lib/copy_saved_objects_to_space/summarize_copy_result.ts:
```diff
@@ -0,0 +1,142 @@
+    // The `savedObject.references` array only includes the direct references. It does not include any references of references.
+    // Therefore, if there are conflicts detected in these transient references, we need to include them here so that they are visible
+    // in the UI as resolvable conflicts.
+    const transientConflicts = conflicts.filter(c => !objectMap.has(`${c.obj.type}:${c.obj.id}`));
```
🙈

legrego commented:
> We also are re-using types and processImportResponse from the saved objects import, which is somewhat awkward. It isn't obvious that internally copy to space is relying upon import/export.

I didn't see it that way - I liked that we weren't duplicating functionality or type definitions for what's effectively a thin wrapper around the import/export feature. It felt to me like we'd be able to see how changes to import/export will impact Copy to Space in the future. Do you have suggestions on how we can improve this? We can divorce the two from a types perspective, but then I feel like it's even less obvious that copy to space relies upon import/export.

> Also, when choosing to "overwrite" a saved object, should we change the test of the "Finish" button? It wasn't immediately obvious to me that clicking "overwrite" wasn't enough to actually overwrite it and that we were just marking it to be retried once we hit "Finish".

Sure, that's a good idea.

legrego commented on x-pack/legacy/plugins/spaces/public/lib/hooks/use_spaces.ts:
```diff
@@ -0,0 +1,28 @@
+
+  useEffect(() => {
+    setIsLoading(true);
+    kfetch({ pathname: '/api/spaces/space', query: { purpose } }).then((response: Space[]) => {
```
Yeah ideally. Passing the `SpacesManager` into the hook would be problematic I think...it's awkward to pass it as a dependency of the hook, because then the hook really isn't doing anything important...it's just calling the `SpacesManager` for you.

I think I'll just remove this hook and replace it with a direct use of `useEffect` within the flyout to take advantage of the `SpacesManager`.

legrego commented on x-pack/legacy/plugins/spaces/public/views/management/components/copy_saved_objects_to_space/selectable_spaces_control.tsx:
```diff
@@ -0,0 +1,83 @@
+export const SelectableSpacesControl = (props: Props) => {
+  const [options, setOptions] = useState<SpaceOption[]>([]);
+
+  useEffect(() => {
```
Good call, this was a holdover from an earlier experiment

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13982/)

legrego commented:
Following a discussion with @kobelb, we decided on the following:
- We will no longer explain when a copy failed due to "missing references"
- We will no longer allow a saved object to be copied without it's references (related objects)

These decisions were made after [discovering shortcomings](https://github.com/elastic/kibana/issues/43873) in the import/export functionality which would have led to a confusing user experience for the Copy to Space feature.

legrego commented:
Also discovered during testing: https://github.com/elastic/kibana/issues/43876

snide commented:
@legrego I'll have a review / PR for you in a couple hours from the design side.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14042/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14095/)
